File: src/leetcode_grind/day_315.rs
// https://leetcode.com/problems/minimum-window-subsequence/description/
pub fn min_window(s1: String, s2: String) -> String {
    let (s1, s2) = (
        s1.chars().collect::<Vec<_>>(),
        s2.chars().collect::<Vec<_>>(),
    );
    let (n, m) = (s1.len(), s2.len());
    let mut dp = vec![vec![10e6 as i32; m + 1]; n + 1];
    dp[0][0] = 0;

    let mut end = 0;
    let mut length = n as i32 + 1;

    for i in 1..=n {
        dp[i][0] = 0;

        for j in 1..=m {
            dp[i][j] = 1 + if s1[i - 1] == s2[j - 1] {
                dp[i - 1][j - 1]
            } else {
                dp[i - 1][j]
            };
        }

        if dp[i][m] < length {
            length = dp[i][m];
            end = i as i32;
        }
    }

    if length > n as i32 {
        "".to_string()
    } else {
        s1[end as usize - length as usize..end as usize]
            .into_iter()
            .collect()
    }
}

File: java/leetcode_grind/Day315.java
package leetcode_grind;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class Day315 {
    static class Solution1 {
        // https://leetcode.com/problems/longest-string-chain/description
        public int longestStrChain(String[] words) {
            var dp = new HashMap<String, Integer>();
            Arrays.sort(words, (a, b) -> a.length() - b.length());
            var longestSubSeqLen = 0;
            for (var word : words) {
                var presentLength = 1;
                for (var i = 0; i < word.length(); i++) {
                    var temp = new StringBuilder(word);
                    temp.deleteCharAt(i);
                    var prevSubSeqLen = dp.getOrDefault(temp.toString(), 0);
                    presentLength = Math.max(presentLength, prevSubSeqLen + 1);
                }
                dp.put(word, presentLength);
                longestSubSeqLen = Math.max(longestSubSeqLen, presentLength);
            }
            return longestSubSeqLen;
        }
    }

    // https://leetcode.com/problems/minimum-window-subsequence/description/
    static class Solution2 {
        public String minWindow(String s1, String s2) {
            var n = s1.length();
            var m = s2.length();

            var answer = "";
            var indices = new HashMap<Character, List<Integer>>();

            for (var i = 0; i < n; i++) {
                var c = s1.charAt(i);
                indices.putIfAbsent(c, new ArrayList<>());
                indices.get(c).add(i);
            }

            var ind = new int[m];
            for (var start = 0; start < n; start++) {
                var prev = start - 1;
                for (var j = 0; j < m; j++) {
                    if (!indices.containsKey(s2.charAt(j))) {
                        return "";
                    }
                    var curIndices = indices.get(s2.charAt(j));
                    while (ind[j] < curIndices.size() && curIndices.get(ind[j]) <= prev) {
                        ind[j]++;
                    }
                    if (ind[j] == curIndices.size()) {
                        return answer;
                    }
                    prev = curIndices.get(ind[j]);
                }

                if (answer.isEmpty() || prev - start + 1 < answer.length()) {
                    answer = s1.substring(start, prev + 1);
                }
            }
            return answer;
        }
    }
}
